### PR TITLE
mpd: 0.20.13 -> 0.21.16

### PR DESCRIPTION
**Pull request by tobim**
###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by tobim** [comment]
The first commit removes the `xSupport` flags from the parameter attribute set and puts the platform specific switches into a single place. The second does the actual update and switch to meson.

**Comment by tobim** [comment]
@GrahamcOfBorg build mpd

**Comment by tobim** [comment]
Added an alternative `mpd-small` attribute for users who want to avoid xorg dependencies.
Closure sizes are:
```
mpd: 352.5M
mpd-small: 60.9M
```

**Comment by tobim** [comment]
maintainer cc: @astsmtl @fuuzetsu @ehmry @fpletz

**Comment by immae** [comment]
This derivation outputs a directory `$(out)/etc/systemd` as is (with `$(out)` not expanded), not critical bug but I think there is a typo somewhere :)

**Comment by immae** [comment]
I suggest this change about above issue and a new flag for better compatibility (I have no opinion whether syslog should be in the list of features or not, but it should definitely be supported)

```patch
118a119
>         ++ [ "-Dsyslog=enabled" ]
122c123
<           "-Dsystemd_system_unit_dir=$(out)/etc/systemd/system";
---
>           "-Dsystemd_system_unit_dir=etc/systemd/system";
```

Also, 0.21.7 is released now

**Comment by tobim** [comment]
Thanks, fixed the path and updated to 0.21.8.
I don't think syslog should be enabled, since the logs are ingested by jounald.

**Comment by immae** [comment]
@tobim: no, the choice is only between syslog and /var/log, so the feature should definitely be switchable in the configuration like the over features available in the derivation

**Comment by immae** [comment]
Sorry, I’ll rephrase: without `"-Dsyslog=enabled"` you can only store logs as a plain file. With the option enabled, you can chose between a file and syslog.

**Comment by Ekleog** [comment]
@GrahamcOfBorg build mpd mpd-small nixosTests.mpd

File: pkgs/servers/mpd/default.nix
Comment (Ekleog) [inline pkgs/servers/mpd/default.nix]: I'd suggest that an error is raised when trying to add an unsupported feature, it would be better than silently ignoring it IMO

File: pkgs/servers/mpd/default.nix
Comment (Ekleog) [inline pkgs/servers/mpd/default.nix]: The hash should likely be kept close to `version` (so, for the time being, to `major` and `minor`), to make upgrading easier.

File: pkgs/servers/mpd/default.nix
Comment (Ekleog) [inline pkgs/servers/mpd/default.nix]: I can't see a reason for the version to be kept in separate `major` and `minor` fields, instead of this just being `version`. AFAIR, @r-ryantm will do upgrades only when the version can be found in a block, so this would be losing an advantage of tooling for no gain that is clear to me.

File: pkgs/servers/mpd/default.nix
Comment (Ekleog) [inline pkgs/servers/mpd/default.nix]: I kind of feel this would be more clear with `mkEnable` and `mkDisable` just be integrated directly as lambdas here, instead of having to go back up in the file to figure the way they're implemented. No strong feelings though.

File: pkgs/servers/mpd/default.nix
Comment (Ekleog) [inline pkgs/servers/mpd/default.nix]: This is `builtins.attrNames`

File: pkgs/servers/mpd/default.nix
Comment (Ekleog) [inline pkgs/servers/mpd/default.nix]: This would be better expressed as `builtins.elem "zeroconf" features_`

File: pkgs/servers/mpd/default.nix
Comment (Ekleog) [inline pkgs/servers/mpd/default.nix]: This refactor looks great to me! It'd be even better if the `run` function was exposed upwards, IMHO, with something like `mpdWithFeatures = run`.

**Review by Ekleog — DISMISSED** [review]
This refactor looks pretty neat to me! A few comments below, though

File: pkgs/servers/mpd/default.nix
Comment (tobim) [reply pkgs/servers/mpd/default.nix]: Makes sense, I actually don't remember why `version` was split like that.

Comment (ryantm) [reply pkgs/servers/mpd/default.nix]: @Ekleog is correct about the version needing to be contiguous.

**Comment by aanderse** [comment]
@Ekleog ping (triage)

**Comment by Ekleog** [comment]
@aanderse I don't particularly know mpd (or even use it myself), just came across this PR while doing triage too. All the comments I could find have been addressed, but I don't feel ready to merge unless someone else reviews it, as it's a pretty big refactor of a derivation for a package I know basically nothing about :)

**Comment by tobim** [comment]
I'm not sure this refactor is still the right approach after the `meson` builder was changed to enable feature auto-detection. I think this can be simplified considerably now.

**Comment by tobim** [comment]
Updated to 0.21.11 and rebased.

**Review by FRidh — COMMENTED** [review]
Why this refactoring? What does it bring? The approach with `fooSupport` is common in Nixpkgs.

**Review by FRidh — CHANGES_REQUESTED** [review]
The maintainers of this expression should agree with this refactoring.

**Comment by tobim** [comment]
> Why this refactoring? What does it bring? The approach with `fooSupport` is common in Nixpkgs.

Have a look at the discussion on this earlier attempt to update this package to meson: https://github.com/NixOS/nixpkgs/pull/51766. The opinion there was that large sets of `enableFoo`-styple options are problematic, and I agree on that part.

> The maintainers of this expression should agree with this refactoring.

This PR has been open for more than 4 months, and none of them commented on it. Maybe I should remove them and add myself?

ping @astsmtl @fuuzetsu @ehmry @fpletz

**Comment by Fuuzetsu** [comment]
>This PR has been open for more than 4 months, and none of them commented on it. Maybe I should remove them and add myself?

You could add yourself, I don't know about removing others.

**Comment by pstch** [comment]
This is my first time trying to help a review by testing a package, so I may have got things wrong, but on my system, this update breaks mpd with the following error :

    exception: Invalid path at line 7:
    exception: nested: not an absolute path: syslog

My MPD configuration file is the following (generated by `home-manager` with default values):

```
music_directory     "/home/pistache/msc"
playlist_directory  "/home/pistache/.local/share/mpd/playlists"
db_file             "/home/pistache/.local/share/mpd/tag_cache"

state_file          "/home/pistache/.local/share/mpd/state"
sticker_file        "/home/pistache/.local/share/mpd/sticker.sql"
log_file            "syslog"

bind_to_address "127.0.0.1"
```

The MPD changelog for 0.21 doesn't mention anything about `log_file` or `syslog`, and the documentation states that `syslog` is an allowed value for `log_file`, so I suppose this is a problem with the build.

**Comment by tobim** [comment]
Thank you for the heads up. I'll try to fix the issue and update to the latest patch level later today.

**Comment by tobim** [comment]
@pstch could you try this PR again with the home-manager module?

**Comment by pstch** [comment]
@tobim I will do that as soon as possible (I'm travelling right now).

Side question: I see you removed the `log_file "syslog"` line, presumably so that MPD logs directly to systemd's journal. MPD is also run as a user service in home-manager, so shouldn't we make it log directly to the journal in home-manager as well (removing `log_file "syslog"` in its MPD config)? This would remove the dependency on syslog, but it may still be a good thing to have.

**Comment by tobim** [comment]
Yes, logging directly to the journal should be done in the user service from home-manager as well. But since home-manager is a separate repository, this cannot be done atomically. It is better to keep the syslog support and update home-manager in a follow-up step. I initially meant to remove `syslog` support but I changed my mind on it. `syslog` support has no extra dependencies so it's there is no cost anyway.

**Comment by pstch** [comment]
OK, makes sense. I can confirm that the updated PR works fine with `home-manager`.

**Comment by doronbehar** [comment]
I've rebased @rycee's `mpd-0.21` branch to master and tested with MPD configured as a module and it works fine. It had to do an update to my music library and it took a while but besides the upgrade was clean. I think `nixos-unstable` users will handle this.

**Comment by doronbehar** [comment]
The rebase wasn't totally trivial, I've pushed the rebased version to my fork: https://github.com/doronbehar/nixpkgs/tree/mpd-0.21

**Comment by tobim** [comment]
> The rebase wasn't totally trivial, I've pushed the rebased version to my fork: https://github.com/doronbehar/nixpkgs/tree/mpd-0.21

Thanks for the heads-up. I rebased this PR on top of nixpkgs-unstable and updated to v0.21.15.

**Review by fpletz — APPROVED** [review]
So I haven't commented so far because I'm not a big fan of changing these feature flags because it breaks peoples' nix expressions. It would be nice if there was a standard we would follow but currently there isn't.

Since you're obviously committed and keen on maintaining mpd - more so than the old maintainers - I see no real reason why we shouldn't merge this now. Thanks a lot for your work on this!

**Comment by arcnmx** [comment]
> It would be nice if there was a standard we would follow but currently there isn't.

As someone who is currently maintaining a branch/patchset that adds a feature to mpd, this would be nice, particularly if it also enabled some way to add to the list via overrides.

**Comment by fpletz** [comment]
Also fixed the mpd NixOS test (which was broken even before this PR): 4d9899ef13aafeb66c3a3d02948966430448ccfe

**Pull request merged**
